### PR TITLE
Refactor import resolver to allow for dynamic loading of modules.

### LINKED ISSUE #1569: Compiler Crashes on Unavailable Imports
### General Summary
Codegen assumes that the imports that it adds into the runtime program representation have been resolved and that they exist. Unfortunately this is not the case.

### Steps to Reproduce

1. Navigate to `Vector_Builder.enso` and change the 3rd line to read `from Standard.Builtins import Array`.
2. Run the database tests.

### Expected Result

A runtime error that mentions the failed resolution of the import.

### Actual Result

A compiler crash.

### Enso Version

```
Enso Compiler and Runtime
Version:    0.2.7-SNAPSHOT
Built with: scala-2.13.5 for GraalVM 21.0.0.2
Built from: wip/ara/project-shadow @ e999457899c5e7b5fdc237a6d03fb2e476d2a66b
Running on: Java HotSpot(TM) 64-Bit Server VM, Oracle Corporation, JDK 11.0.10+8-LTS-jvmci-21.0-b06
            Mac OS X 10.16 (x86_64)
```

### LINKED ISSUE #1768: Refactor Import Resolution To Use a Package Repository
### Summary

We want to introduce a concept of a `PackageRepository` which resolves library names to library paths and downloads any missing libraries.

This part consists of creating a dummy `PackageRepository` that is still using the old import logic but most importantly refactoring the `ImportResolver` to use the `PackageRepository` interface, allowing for future migration.

### Value

- First step towards dynamically loading libraries.

### Specification

- [ ] Move the Context initialization logic that is related to preloading libraries into a legacy implementation of `PackageRepository`.
- [ ] Make the `ImportResolver` use a `PackageRepository` for requesting `Module`s.
- [ ] Add a callback that is called when `PackageRepository` encounters a new library (that was not previously loaded) which notifies the language-server (if it is connected) that a new content root may be added.
- [ ] Make sure that module update logic also uses the `PackageRepository`.
  - [ ] That includes all relevant references within the `Context`, for example the `packages` list should also be updated to rely on the `PackageRepository`.
- [ ] Update the `TopLevelScope` to refer to the `PackageRepository`.
  - [ ] In particular, it may be good to rename operations like `getModules` to `getLoadedModules`.

### Acceptance Criteria & Test Cases

- The above specification is completed.
- [ ] The import resolution logic and the callback have been tested.

### PR DESCRIPTION
**Pull request by kustosz**
### Pull Request Description
Fixes #1569 
Closes #1768 

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

**Review by iamrecursion — APPROVED** [review]

File: engine/runtime/src/main/scala/org/enso/compiler/PackageRepository.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  /** Informs the repository that it should populate the top scope with modules
+    * belonging to a given package.
+    *
+    * @param namespace the namespace of the package.
+    * @param name the package name.
+    * @return `true` if the package was already loaded or successfully
+    *         downloaded. `false` otherwise.
+    */
+  def ensurePackageIsLoaded(namespace: String, name: String): Boolean
```
Comment (radeusgd) [inline engine/runtime/src/main/scala/org/enso/compiler/PackageRepository.scala]: I would suggest extending this signature a bit, although I guess it is ok if it happens in future PRs.

I think we should here somehow differentiate various kinds of errors - for display purposes there should be a different error saying "package is not defined in the edition that you are using" versus "package could not be downloaded due to network problems" or "could not load the package due to insufficient filesystem permissions", as these are completely different error, so thay cannot be merged into a single `false` return value.

**Review by radeusgd — APPROVED** [review]
Looks good to me.

I think error handling of failed module load should be more granular as indicated below, but I guess that can be further extended in future PRs.
Maybe this PR could at least change the return type of `ensurePackageIsLoaded` into an `Either[PackageLoadingError, Unit]` with only one variant defined for `PackageLoadingError` which can then be later extended.

**Comment by radeusgd** [comment]
After having a second look at the spec I have some doubts if everything is done - technically I guess all features are there, but I think there are things that could be done in this PR instead of the next one.

For example, the `getPackages` method of the context may be misleading as it only returns the preloaded packages - but as it is only used in the temporary PackageRepository, maybe its logic could be moved into the PackageRepository itself?
Currently the PackageRepository delegates a lot back to the Context, so they are actually quite tightly coupled while the whole purpose of this task was to delegate as much as possible of module handling to the PackageRepository, so that the next task will require less modifications.

I'm also not sure if we should keep `modules` in Context. I mean I think this is a correct solution, and I appreciate that we only explicitly try to load new packages by calling the `ensurePackageIsLoaded` method. On the other hand, I feel like having modules there may result in some other code, perhaps added later, getting modules directly from this field and not calling `ensurePackageIsLoaded`, which could result in forgetting to preload some packages. But maybe it is indeed better to ask to load packages explicitly, rather then implicitly with each `getModuleByName` request? I'm not sure, just wanted to discuss the alternatives. I would however suggest to rename these methods to indicate that they only refer to loaded modules and will not work if a package was not yet loaded.

File: engine/runtime/src/main/scala/org/enso/compiler/PackageRepository.scala (hunk 1 above)
Comment (kustosz) [reply engine/runtime/src/main/scala/org/enso/compiler/PackageRepository.scala]: Fair!

**Comment by kustosz** [comment]
@radeusgd honestly this is just an unblocker for "things in compiler use this, it is obvious where it should be replaced in case the API changes", such that further changes to the packaging logic don't need to propagate far into the compiler. Regarding moving the preloading logic into package repo – I decided against it on the basis of "what's the point moving code that will be removed by the next task?". There will be no preloading, so this whole logic will be removed. Regarding "module access should be mediated through the package repo" – interesting thought, it does have some problems, but may be more elegant. I'll think about it.

**Comment by radeusgd** [comment]
> Regarding moving the preloading logic into package repo – I decided against it on the basis of "what's the point moving code that will be removed by the next task?"

Fair

> Regarding "module access should be mediated through the package repo" – interesting thought, it does have some problems, but may be more elegant. I'll think about it.

That was my initial idea, I think that it reduces the chance of erroneously forgetting to call `ensurePackageIsLoaded`. But also being explicit about loading also may have some advantages, so yeah it would be good to think through which one will be better overall.

**Pull request merged**